treq/auth.py

```diff
@@ -26,6 +84,228 @@ def request(self, method, uri, headers=None, bodyProducer=None):
+                "%s:%s" % (ha1, "%s:%s" % (ha2, nonce))
+            )
+        else:
+            noncebit = "%s:%s:%s:%s:%s" % (
+                nonce, ncvalue, cnonce.encode('utf-8'), 'auth', ha2
+            )
+            response_digest = digest_hash_func("%s:%s" % (ha1, noncebit))
+
+        hb = 'username="%s", realm="%s", nonce="%s", uri="%s", response="%s"' % (
+            self._username, realm, nonce, actual_path, response_digest
+        )
+        if opaque:
+            hb += ', opaque="%s"' % opaque
+        if original_algo:
+            hb += ', algorithm="%s"' % original_algo
+        if qop:
+            hb += ', qop="auth", nc=%s, cnonce="%s"' % (ncvalue, cnonce)
+        if not cached:
+            cache_params = {
+                'agent': agent,
+                'path': path,
+                'method': method,
+                'cached': cached,
+                'nonce': nonce,
+                'realm': realm,
+                'qop': qop,
+                'algorithm': algorithm,
+                'opaque': opaque
+            }
+            agent.digest_auth_cache[(method, path)] = {
+                'p': cache_params,
+                'c': 1
+            }
+        return 'Digest %s' % hb
+
+    def _on_401_response(self, www_authenticate_response, method, uri, headers,
+                         bodyProducer):
+        """
+        Handle the server`s 401 response, that is capable with authentication
+            headers, build the Authorization header
+        for
+        :param www_authenticate_response: t.w.client.Response object
+        :param method: HTTP method to be used to perform the request
+        :param uri: URI to be used
+        :param headers: Additional headers to be sent with the request,
+            instead of "Authorization" header
+        :param bodyProducer: IBodyProducer implementer instance that would be
+            used to fetch the response body
+        :return:
+        """
+        assert www_authenticate_response.code == 401, \
+            """Got invalid pre-authentication response code, probably URL
+            does not support Digest auth
+        """
+        www_authenticate_header_string = www_authenticate_response.\
+            headers._rawHeaders.get('www-authenticate', [''])[0]
+        digest_authentication_params = parse_dict_header(
+            _DIGEST_HEADER_PREFIX_REGEXP.sub(
+                '', www_authenticate_header_string, count=1
+            )
+        )
+        if digest_authentication_params.get('qop', None) is not None and \
+            digest_authentication_params['qop'] != 'auth' and \
+                'auth' not in digest_authentication_params['qop'].split(','):
+            # We support only "auth" QoP as defined in rfc-2617 or rfc-2069
+            raise UnknownQopForDigestAuth(digest_authentication_params['qop'])
+        digest_authentication_header = self._build_digest_authentication_header(
+            self,
```

Comment:
This self (and the agent argument in `_build_digest_authentication_header` are redundant.